### PR TITLE
refactor(ethereum-client): Eth client update

### PR DESCRIPTION
**Pull request by koinov**
ethereum client update, Contracts with generics, EthApi, tests with alloy

### CODE REVIEW

File: ethereum/client/src/abi.rs [hunk 1]
```diff
@@ -31,12 +31,65 @@ sol! {
...
+
+
 }
 
+#[cfg(test)]
+sol!(
+    #[sol(rpc)]
+    ProxyContract,
+    "../out/ProxyContract.sol/ProxyContract.json"
+);
```
Comment (mertwole) [inline ethereum/client/src/abi.rs:45]:
```suggestion
#[cfg(test)]
mod ... {
sol!(
    #[sol(rpc)]
    ProxyContract,
    "../out/ProxyContract.sol/ProxyContract.json"
);

...
}
```

File: ethereum/client/src/abi.rs [hunk 2]
```diff
@@ -31,12 +31,65 @@ sol! {
...
+
+#[cfg(test)]
+sol!(
+    #[sol(rpc)]
+    ERC20Mock,
+    "../out/ERC20Mock.sol/ERC20Mock.json"
```
Comment (mertwole) [inline ethereum/client/src/abi.rs:51]: Can you generate these `.json`s in the `build.rs` script?

File: ethereum/client/src/lib.rs [hunk 3]
```diff
@@ -29,34 +32,37 @@ use crate::{
...
+            alloy::providers::fillers::ChainIdFiller,
         >,
+        WalletFiller<alloy::network::EthereumWallet>,
     >,
+    RootProvider<alloy::transports::http::Http<alloy::transports::http::Client>>,
+    alloy::transports::http::Http<alloy::transports::http::Client>,
+    Ethereum,
 >;
 
+pub use alloy::primitives::TxHash;
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:56]: Move it to the other `use`s

File: ethereum/client/src/lib.rs [hunk 4]
```diff
@@ -29,34 +32,37 @@ use crate::{
...
+
 #[derive(Clone)]
-pub struct Contracts {
-    signer: Wallet<SigningKey>,
-    provider: Arc<ProviderType>,
-    message_queue_instance: IMessageQueueInstance<Ethereum, BoxTransport, Arc<ProviderType>>,
-    relayer_instance: IRelayerInstance<Ethereum, BoxTransport, Arc<ProviderType>>,
+pub struct Contracts<P, T, N> {
+    //signer: PrivateKeySigner,
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:60]:
```suggestion
```

File: ethereum/client/src/lib.rs [hunk 5]
```diff
@@ -29,34 +32,37 @@ use crate::{
...
+    provider: P,
+    message_queue_instance: IMessageQueueInstance<T, P, N>,
+    relayer_instance: IRelayerInstance<T, P, N>,
+    _t: PhantomData<T>,
+    _n: PhantomData<N>,
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:65]:
```suggestion
    _phantom: PhantomData<(T, N)>,
```

File: ethereum/client/src/lib.rs [hunk 6]
```diff
@@ -503,14 +690,16 @@ mod tests {
         println!("Proof : {:?}", proof);
     }
 
+    //TODO Check
+    /*
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:694]: Don't leave commented out code

File: ethereum/client/src/lib.rs [hunk 7]
```diff
@@ -358,72 +430,189 @@ impl Contracts {
...
                 "block_number" : 273,
                 "merkle_root" : "0xa25559d02a45bf58afd5344964269d38e947a432c1097c342f937a4ad052a683"
             }"#;
         proof_json.to_string()
     }
 
+    #[tokio::test]
+    async fn test_deploy() -> Result<(), Error> {
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:538]: What's this test purpose?

File: ethereum/client/src/lib.rs [hunk 8]
```diff
@@ -358,72 +430,189 @@ impl Contracts {
 
 #[cfg(test)]
 mod tests {
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:432]: Please move tests to the separate file

File: ethereum/client/src/lib.rs [hunk 9]
```diff
@@ -358,72 +430,189 @@ impl Contracts {
...
-    ) -> (HttpProvider<Ethereum>, AnvilInstance) {
-        let anvil = f(Anvil::new()).try_spawn().expect("could not spawn anvil");
-        (anvil_http_provider(&anvil), anvil)
-    }
+    use super::*;
 
-    #[allow(unused, unreachable_pub)]
-    pub fn anvil_http_provider(anvil: &AnvilInstance) -> HttpProvider<Ethereum> {
-        http_provider(&anvil.endpoint())
+    fn build_contracts<P, T>(provider: P) -> Result<Contracts<P, T, Ethereum>, Error>
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:449]: Why not to take addresses from `DeploymentEnv` after `deploy()`?

File: ethereum/client/src/lib.rs [hunk 10]
```diff
@@ -358,72 +430,189 @@ impl Contracts {
...
-    use alloy_node_bindings::{Anvil, AnvilInstance};
-    use alloy_primitives::keccak256;
-    use alloy_provider::HttpProvider;
-    use alloy_transport_http::Http;
+
+    use alloy::hex;
+    use alloy::node_bindings::{Anvil, AnvilInstance};
+    use alloy::primitives::keccak256;
+    use alloy::providers::ProviderBuilder;
+    use alloy::transports::http::{Client, Http};
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:438]: Common prefixes in deps

File: ethereum/client/src/lib.rs [hunk 11]
```diff
@@ -281,7 +352,7 @@ impl Contracts {
     }
 
     pub async fn is_message_processed(&self, nonce_le: [u8; 32]) -> Result<bool, Error> {
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:354]:
```suggestion
    pub async fn is_message_processed(&self, nonce: [u8; 32]) -> Result<bool, Error> {
```

File: ethereum/client/src/lib.rs [hunk 12]
```diff
@@ -212,54 +299,38 @@ impl Contracts {
...
-                    .gas_price(U256::from(2_000_000_000_000u128))
-                    .send()
-                    .await
-                {
-                    Ok(pending_tx) => Ok(*pending_tx.tx_hash()),
-                    Err(e) => {
-                        log::error!("Sending error: {e:?}");
-                        Err(Error::ErrorSendingTransaction(e))
-                    }
+            Ok(_gas_used) => match call.send().await {
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:327]: What the gas price will be used here?

File: ethereum/client/src/lib.rs [hunk 13]
```diff
@@ -158,9 +244,10 @@ impl Contracts {
         }
     }
 
-    pub async fn provide_merkle_root_json(&self, json_string: &str) -> Result<TxHash, Error> {
+    pub async fn provide_merkle_root_json(&self, json_string: &str) -> Result<B256, Error> {
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:247]: As I see, it used only in tests, so why not to remove it and use `provide_merkle_root` everywhere instead?

File: ethereum/client/src/lib.rs [hunk 14]
```diff
@@ -73,51 +79,133 @@ pub enum TxStatus {
     Failed,
 }
 
-impl Contracts {
+#[derive(Clone)]
+pub struct EthApi {
+    contracts: Contracts<ProviderType, Http<Client>, Ethereum>,
+}
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:85]:
```suggestion
pub struct EthApi(Contracts<ProviderType, Http<Client>, Ethereum>)
```

Comment (mertwole) [reply ethereum/client/src/lib.rs:85]: And implement `Deref` for `EthApi` to not duplicate `Contracts` API

File: ethereum/client/src/lib.rs [hunk 15]
```diff
@@ -29,34 +32,37 @@ use crate::{
...
+                alloy::providers::fillers::JoinFill<alloy::providers::Identity, GasFiller>,
+                NonceFiller,
+            >,
+            alloy::providers::fillers::ChainIdFiller,
         >,
+        WalletFiller<alloy::network::EthereumWallet>,
     >,
+    RootProvider<alloy::transports::http::Http<alloy::transports::http::Client>>,
+    alloy::transports::http::Http<alloy::transports::http::Client>,
+    Ethereum,
```
Comment (mertwole) [inline ethereum/client/src/lib.rs:53]: Add usings for `alloy` types

**Review by mertwole — CHANGES_REQUESTED** [review]
Also fix clippy warnings

File: ethereum/client/src/abi.rs (hunk 2 above)
Comment (koinov) [reply ethereum/client/src/abi.rs:51]: no, but can do from Makefile

File: ethereum/client/src/lib.rs (hunk 3 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:56]: this is pub use used outside the crate.

File: ethereum/client/src/lib.rs (hunk 4 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:60]: done

File: ethereum/client/src/lib.rs (hunk 5 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:65]: done

File: ethereum/client/src/lib.rs (hunk 8 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:432]: Tried to find example in this project without success.

File: ethereum/client/src/lib.rs (hunk 10 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:438]: Done

File: ethereum/client/src/lib.rs (hunk 11 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:354]: Done

File: ethereum/client/src/lib.rs (hunk 14 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:85]: Done, but Deref brings out ethereum types those are not really necessary

File: ethereum/client/src/lib.rs (hunk 15 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:53]: Done

File: ethereum/client/src/lib.rs (hunk 14 above)
Comment (mertwole) [reply ethereum/client/src/lib.rs:85]: Discussed in DM

File: ethereum/client/src/abi.rs (hunk 1 above)
Comment (mertwole) [reply ethereum/client/src/abi.rs:45]: Not relevant

File: ethereum/client/src/lib.rs (hunk 3 above)
Comment (mertwole) [reply ethereum/client/src/lib.rs:56]: I mean move it to the line `#34`, after `use crate::...`

Comment (koinov) [reply ethereum/client/src/lib.rs:56]: Done

File: ethereum/client/src/lib.rs (hunk 6 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:694]: Done

File: ethereum/client/src/lib.rs (hunk 7 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:538]: To check check deployed contracts addresses

File: ethereum/client/src/lib.rs (hunk 8 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:432]: Done

File: ethereum/client/src/lib.rs (hunk 12 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:327]: Gas price determined by GasFiller in Alloy  Provider that is added by calling with_recommended_fillers()

File: ethereum/client/src/lib.rs (hunk 13 above)
Comment (koinov) [reply ethereum/client/src/lib.rs:247]: Removed, part of code moved to tests.rs

File: ethereum/client/src/abi.rs (hunk 2 above)
Comment (koinov) [reply ethereum/client/src/abi.rs]: Added some decorators

File: ethereum/client/src/lib.rs (hunk 9 above)
Comment (koinov) [reply ethereum/client/src/lib.rs]: Done

File: ethereum/client/src/abi.rs [hunk 16]
```diff
@@ -1,9 +1,10 @@
-use alloy_sol_types::sol;
+#![allow(dead_code)]
```
Comment (mertwole) [inline ethereum/client/src/abi.rs:1]: Unreasonable

File: ethereum/client/src/abi.rs [hunk 17]
```diff
@@ -31,12 +32,16 @@ sol! {
 
         function getBlockNumber(bytes32 merkle_root) external view returns (uint256);
     }
+
+
+
+
```
Comment (mertwole) [inline ethereum/client/src/abi.rs:38]:
```suggestion
```

File: ethereum/client/src/tests.rs [hunk 18]
```diff
@@ -0,0 +1,281 @@
+use alloy::sol;
+
+#[cfg(test)]
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:3]: It's already under `#[cfg(test)]` in `lib.rs`

File: ethereum/client/src/tests.rs [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+);
+
+#[cfg(test)]
+sol!(
+    #[sol(rpc)]
+    ERC20Treasury,
+    "../out/ERC20Treasury.sol/ERC20Treasury.json"
+);
+#[cfg(test)]
+mod test {
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:52]: You don't need this `mod` anymore

File: ethereum/client/src/tests.rs [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+                "- Block : {} Merkle Root : {} TxHash : {}",
+                merkle_root_entry.block_number,
+                merkle_root_entry.merkle_root,
+                merkle_root_entry.tx_hash
+            )
+        }
+    }
+
+    #[tokio::test]
+    async fn verify_merkle_proof() {
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:263]: This test doesn't assert anything

File: ethereum/client/src/tests.rs [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+        assert_ne!(logs.len(), 0);
+
+        for merkle_root_entry in logs.iter() {
+            println!(
+                "- Block : {} Merkle Root : {} TxHash : {}",
+                merkle_root_entry.block_number,
+                merkle_root_entry.merkle_root,
+                merkle_root_entry.tx_hash
+            )
+        }
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:259]: Why do you need to print it?

File: ethereum/client/src/tests.rs [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+            message_queue: *message_queue.address(),
+            relayer: *relayer.address(),
+            erc20_treasury: *erc20_treasury.address(),
+            message_queue_proxy: *message_queue_proxy.address(),
+            relayer_proxy: *relayer_proxy.address(),
+            erc20_treasury_proxy: *erc20_treasury_proxy.address(),
+        })
+    }
+
+    fn build_merkle_proof_json() -> String {
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:170]: Just construct `BlockMerkleRootProof` directly without parsing and deserializing it

File: ethereum/client/src/tests.rs [hunk 23]
```diff
@@ -0,0 +1,281 @@
+use alloy::sol;
+
+#[cfg(test)]
+sol!(
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:4]: Can you squash these `sol!` declarations into one?

**Review by mertwole — CHANGES_REQUESTED** [review]

File: ethereum/client/src/abi.rs (hunk 17 above)
Comment (koinov) [reply ethereum/client/src/abi.rs:42]: Done

File: ethereum/client/src/abi.rs (hunk 16 above)
Comment (koinov) [reply ethereum/client/src/abi.rs:1]: Done

File: ethereum/client/src/tests.rs (hunk 19 above)
Comment (koinov) [reply ethereum/client/src/tests.rs:52]: Done

File: ethereum/client/src/tests.rs (hunk 21 above)
Comment (koinov) [reply ethereum/client/src/tests.rs:259]: Removed

File: ethereum/client/src/tests.rs (hunk 20 above)
Comment (koinov) [reply ethereum/client/src/tests.rs:263]: Fixed

File: ethereum/client/src/lib.rs (hunk 7 above)
Comment (koinov) [reply ethereum/client/src/lib.rs]: removed

File: ethereum/client/src/tests.rs (hunk 18 above)
Comment (koinov) [reply ethereum/client/src/tests.rs]: fixed

File: ethereum/client/src/tests.rs (hunk 22 above)
Comment (koinov) [reply ethereum/client/src/tests.rs]: done

File: ethereum/client/src/convert.rs
Comment (mertwole) [inline ethereum/client/src/convert.rs:1]: Remove it completely

File: ethereum/client/src/tests.rs [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+    "../out/MessageQueue.sol/MessageQueue.json"
+);
+
+sol!(
+    #[sol(rpc)]
+    Verifier,
+    "../out/Verifier.sol/Verifier.json"
+);
+
+#[cfg(test)]
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:33]: You don't need `#[cfg(test)]` anywhere in this file

File: ethereum/client/src/tests.rs [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+    transports::Transport,
+};
+
+use binary_merkle_tree::{merkle_proof, merkle_root, Leaf, MerkleProof};
+use primitive_types::H256;
+use sp_core::KeccakHasher;
+
+use crate::{abi::ContentMessage, Contracts, Error};
+
+use serde::{Deserialize, Serialize};
```
Comment (mertwole) [inline ethereum/client/src/tests.rs:60]: Please move usings to the top

File: ethereum/client/src/convert.rs
Comment (koinov) [reply ethereum/client/src/convert.rs:1]: Done

File: ethereum/client/src/tests.rs (hunk 25 above)
Comment (koinov) [reply ethereum/client/src/tests.rs]: Done

File: ethereum/client/src/tests.rs (hunk 24 above)
Comment (koinov) [reply ethereum/client/src/tests.rs]: Fixed

**Pull request merged**